### PR TITLE
build providers: recreate instance if base changed

### PR DESCRIPTION
**Pull request by cmatsuoka**
Once a build provider is provisioned, changes in project base are
ignored and the existing VM is used for new builds. To ensure base
and provider contents are consistent, we test base values at the
start of each build and force instance recreation if base has
changed.

LP: #1794506

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [ ] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/internal/build_providers/_base_provider.py [hunk 1]
```diff
@@ -240,7 +246,16 @@ def launch_instance(self) -> None:
             # what is on the host
             self._setup_snapcraft()
 
+    def _verify_base_change(self) -> None:
+        info = self._load_info()
+        print(info, file=open("/tmp/lala", "w"))
```
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply snapcraft/internal/build_providers/_base_provider.py]: Debug code removed. It reminds me when the developer of a network card driver left a debug printk("hello") in the code and people started asking what was that in their kernel logs.

File: snapcraft/internal/build_providers/_base_provider.py [hunk 2]
```diff
@@ -240,7 +246,16 @@ def launch_instance(self) -> None:
             # what is on the host
             self._setup_snapcraft()
 
+    def _verify_base_change(self) -> None:
```
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply snapcraft/internal/build_providers/_base_provider.py]: I had this section of code inlilne in caller, but ended up with a separate function to make it testable separately. I'll refactor that to something cleaner.

File: snapcraft/internal/build_providers/_base_provider.py [hunk 3]
```diff
@@ -246,10 +246,15 @@ def launch_instance(self) -> None:
             # what is on the host
             self._setup_snapcraft()
 
-    def _verify_base_change(self) -> None:
+    def _ensure_base(self) -> None:
         info = self._load_info()
         provider_base = info["base"] if "base" in info else None
         if self._base_has_changed(self.project.info.base, provider_base):
+            logger.warning(
```
Comment (sergiusens) [inline snapcraft/internal/build_providers/_base_provider.py]: can you use `self.echoer.warn` here instead?

File: snapcraft/internal/build_providers/_base_provider.py [hunk 4]
```diff
@@ -240,7 +246,16 @@ def launch_instance(self) -> None:
             # what is on the host
             self._setup_snapcraft()
 
+    def _verify_base_change(self) -> None:
+        info = self._load_info()
+        print(info, file=open("/tmp/lala", "w"))
+        provider_base = info["base"] if "base" in info else None
+        if self._base_has_changed(self.project.info.base, provider_base):
+            self.clean_project()
```
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply snapcraft/internal/build_providers/_base_provider.py]: Added a message to warn the user.

File: snapcraft/internal/build_providers/_base_provider.py (hunk 3 above)
Comment (cmatsuoka) [reply snapcraft/internal/build_providers/_base_provider.py]: Changed call to `self.echoer.warning`.

Comment (sergiusens) [reply snapcraft/internal/build_providers/_base_provider.py]: hmm, I am not seeing this.

Comment (cmatsuoka) [reply snapcraft/internal/build_providers/_base_provider.py]: Yep, sorry for the confusion. Merged commits from master and a rebase leftover made me think it was fixed elsewhere. It's there now.

**Pull request merged**